### DISCUSSION TITLE
ot-daemon SPI race condition

### DISCUSSION THREAD

**Discussion by kimotine** [openthread/openthread#8698] [Q&A]
Hello,

SPI_INT GPIO is used by the RPC to notify the host (ot-daemon) that it has some data to transmit.
By default, SPI_INT GPIO value is 1. RCP sets SPI_INT GPIO value to 0 in order to notify the host.
Ot-daemon must check the falling edge of this GPIO to initiate a SPI transmission to get the value sent by the RPC.

The mechanism used by ot-daemon the detect the falling edge of SPI_INT involves:
-	**CheckInterrupt()** function reads SPI_INT GPIO value
-	a **FDSET** is called to add the file descriptor matching the SPI_INT GPIO to the READ set
-	use **select()** on the READ set to wait for the falling edge on SPI_INT GPIO. (select() will poll on event sent by the ioctl)
A timeout is set in the select() function in order to stop waiting if ever the SPI_INT GPIO has not changed.
If the SPI_INT GPIO is already at low level when **CheckInterrupt()** is called, FDSET is not necessary and timeout is set to 0. Then **select()** will return immediately. (this is Case 2 in following diagram)

Following figure presents possible behaviors:
![analysis](https://user-images.githubusercontent.com/123661546/214904856-83bebbfe-2db8-4f21-ad0e-1b3a5e924f70.png)

### Problem:
In Case 3, A race condition occurs when SPI_INT GPIO changes between **CheckInterrupt()** and **FD_SET**. 
The timeout is not set to 0 because CheckInterrupt() indicates that SPI_INT GPIO is 1. 
**select()** won’t detect SPI_INT GPIO change and will block until the end of timeout although SPI_INT GPIO is low.

### Fix proposal:
SPI_INT GPIO value must be checked again after FD_SET (with CheckInterrupt) and the timeout set to 0 if SPI_INT value is 0.

### Note:
In ot-daemon implementation, 
CheckInterrupt and FD_SET are called in function **SpiInterface::UpdateFdSet** (spi_nterface.cpp)
Select is called in **otSysMainloopPoll** (system.cpp)

**Comment by abtink** [comment]
Thanks @kimotine for reporting this and detailed description.

I am not sure if this is an issue.

I expect anytime  gpio event is triggered the corresponding `mIntGpioValueFd` will become readable (will have stuff to read).

So I think case 3 should work fine:
- The `FD_SET()` does not start the monitoring  of `fd` (it is macro which adds the given `fd` to a bitmask `fd_set`).
- In case 3, when `select()` is called, it should return immediately (as the `fd` is already readable).
- And upon  return from `select`, the `readFdSet` should contain `mIntGpioValueFd` (to indicate it has stuff to read).
- Then in `Process()` we check this and read the data which clears it (so we  are ready for any next INT):
```c++
    if (FD_ISSET(mIntGpioValueFd, aReadFdSet))
    {
        struct gpioevent_data event;

        otLogDebgPlat("Process(): Interrupt.");

        // Read event data to clear interrupt.
        VerifyOrDie(read(mIntGpioValueFd, &event, sizeof(event)) != -1, OT_EXIT_ERROR_ERRNO);
    }
```

**Comment by kimotine** [comment]
Thanks @abtink for your feed back.

I have seen case 3 behavior during experimentation.
The impact of this case is a SPI latency that can be very long (more than 500ms) although the transmission must be handled immediately.
I have made the modification I suggest and the SPI latency disappeared.

The modification is:

```
        if (CheckInterrupt())
        {
            // Interrupt pin is asserted, set the timeout to be 0.
            timeout.tv_sec  = 0;
            timeout.tv_usec = 0;
            otLogDebgPlat("UpdateFdSet(): Interrupt.");
        }
        else
        {
            // The interrupt pin was not asserted, so we wait for the interrupt pin to be asserted by adding it to the
            // read set.
            FD_SET(mIntGpioValueFd, &aReadFdSet);

            if (CheckInterrupt())
            {
                // Interrupt pin was not asserted before FD_SET
                // but it is after.
                // set the timeout to zero else the select will not detect the falling edge
                // and will wait until the end of the timeout
                timeout.tv_sec  = 0;
                timeout.tv_usec = 0;
                otLogDebgPlat("UpdateFdSet(): Interrupt.");
            }
        }
```

Actually, Case 3 is very close to Case 2, because in both cases, SPI_INT is set to 0 before FD_SET.
I assume the timeout is set to 0 in Case 2, else the select does not return immediately.

**Reply by abtink** [reply]
Here is the behavior I expect:

- Whenever there is a transition from HIGH to LOW on `SPI_INT` pin, its related fd `mIntGpioValueFd` should have an event to `read`.
- This should be tracked all the time, independent of when/if we call `FD_SET()`.
- `FD_SET()` call does not start this process (of monitoring the pin and updating its related fd).
- So calling `CheckInterrupt()` after `FD_SET()` should not matter or help.
- In case 3:
   - We check `CheckInterrupt()`, PIN is is HIGH so we proceed to `else` block
   - Say now GPIO pin transitions to LOW, at this point the `mIntGpioValueFd` should have the event to `read`.
   - We then we add the GPIO fd to the "read fd-set" -> `FD_SET(mIntGpioValueFd, &aReadFdSet)`  - This call should not change the state/content in `mIntGpioValueFd`.
   - We then call `select()` which now should return immediately since the `mIntGpioValueFd` which is included in the "read fd-set" has data for us to read.

Anything I may have missed above or you disagree with?

If you notice delay behavior, there may be another issue but I am not sure if this is root cause. There may some other situation.

**Reply by kimotine** [reply]
Thank you for your comment.
I understand your statement.
In that case, I am surprised it is necessary to set the timeout to 0 in case 2.
Because, in case 2, whatever the value of timeout, the select should return immediately as you mentioned.
What do you think?

I will double check the behavior I have observed.

**Reply by abtink** [reply]
Thanks @kimotine. If you see delayed behavior, it'd be great to figure out the root cause and fix it.

About case 2, I agree we should not need the `CheckInterrupt()` and set timeout to 0 as well.

I think the `CheckInterrupt()` and setting timeout to zero is only used to handle an edge-case on when we start and the very first time we run through the loop. If the GPIO pin is already `LOW` when we start. when we set up its related `mIntGpioValueFd` there will be no transition and therefore no event to read in the `fd`. So we cannot use `select()` in the very first loop to detect this, so we have this direct call to `CheckInterrupt()` and setting of timeout to zero. @zhanglongxia do you happen to remember?

**Reply by abtink** [reply]
I am reading the code and thinking of different cases. I can see this situation below which is not idle but I cannot see it causing any issue. Basically we may run through the loop one more time after handling the GPIO INT already:
- We have case 2 (GPIO INT is already `LOW` and we have an event to read in `mIntGpioValueFd`):
- In the loop, in `UpdateFdSet()` we check and set the timeout to zero but don't add `mIntGpioValueFd` to the "read fd-set".
- `select()` returns quickly, we go to `Process()`.
- But since `mIntGpioValueFd` is not in fd-set, we won't read from it and event remains in the `fd`.
- We do `PushPullSpi()` and peripheral changes GPIO INT to `HIGH` (intterupt is handled).
- The next round we go through the loop, we add `mIntGpioValueFd` to the read fd-set
- Since it still has the unread event, `select()` returns quickly.
- Now `Process()` will read the event from `fd`.

It may be better to change the code to always add `mIntGpioValueFd` to the read-fd set and not just from `else` block of `if (CheckInterrupt())`.

**Reply by kimotine** [reply]
I will do some traces in the coming days (I can spy the SPI GPIO and SPI_INT lines with Saleae logical analyser).
Actually, with this setup, I can clearly see that a delay between SPI_INT falling edge and the SPI communication occurs sometimes.
The solution I mentioned initially fix the issue. That's why I was thinking about a race condition.

But I understand your explanation about `select` , so I will look at it again and let you know.

**Reply by bukepo** [reply]
> Thanks @kimotine. If you see delayed behavior, it'd be great to figure out the root cause and fix it.
> 
> About case 2, I agree we should not need the `CheckInterrupt()` and set timeout to 0 as well.
> 
> I think the `CheckInterrupt()` and setting timeout to zero is only used to handle an edge-case on when we start and the very first time we run through the loop. If the GPIO pin is already `LOW` when we start. when we set up its related `mIntGpioValueFd` there will be no transition and therefore no event to read in the `fd`. So we cannot use `select()` in the very first loop to detect this, so we have this direct call to `CheckInterrupt()` and setting of timeout to zero. @zhanglongxia do you happen to remember?

In addition, I think setting timeout to 0 is can handle some other cases as well:
* RCP keeps SPI_INT low when there are pending frames.
* The interrupt is lost somehow even RCP toggled the pin.

**Reply by zhanglongxia** [reply]
The SPI interface supports the [polling mode](https://github.com/openthread/openthread/blob/main/src/posix/platform/spi_interface.cpp#L179), RCP doesn't need the SPI INT pin in this mode. The function `CheckInterrupt()` is also used to process this case.

**Reply by abtink** [reply]
Thanks @zhanglongxia. Yes. we also have the polling mode but in the polling case,  `mIntGpioValueFd` should remain as `-1` (we don't have a GPIO pin given to us, so we never call `InitIntPin()` and never set/init `mIntGpioValueFd`).

For the polling case this whole `if (mIntGpioValueFd >= 0)` block in `UpdateFdSet()` (which contains the `CheckInterrupt()`) is skipped and we update the `timeout` in `else` to `pollingTimeout`:
https://github.com/openthread/openthread/blob/c28bb194d41d377d1f42ffe50a3916d1545d135c/src/posix/platform/spi_interface.cpp#L641-L666

------

The question is why do we do `CheckInterrupt()` within the `if` block above and not rely on `mIntGpioValueFd` to detect it.
I cannot see any other reason than handling the edge-case on the very first run through the loop (from https://github.com/openthread/openthread/discussions/8698#discussioncomment-4821202).
Do you recall or see any other reason?

**Reply by zhanglongxia** [reply]
> The question is why do we do `CheckInterrupt()` within the `if` block above and not rely on `mIntGpioValueFd` to detect it. I cannot see any other reason than handling the edge-case on the very first run through the loop (from [#8698 (reply in thread)](https://github.com/openthread/openthread/discussions/8698#discussioncomment-4821202)). Do you recall or see any other reason?

I think the main reason is like @bukepo said in [comment](https://github.com/openthread/openthread/discussions/8698#discussioncomment-4825342). The `CheckInterrupt()` is used to process some corner case, such as the interrupt is missed.

**Comment by kimotine** [comment]
Hi @abtink,

After checking the code and adding some traces, I think you miss the `FD_ZERO `at the begin of the loop in main.c

The main loop sequence is:

1. `FD_ZERO(&mainloop.mReadFdSet)`
2. `otSysMainloopUpdate(instance, &mainloop)`
2.1 `CheckInterrupt()`
2.2 timer = 0
2.3 or `FD_SET(mIntGpioValueFd, &aReadFdSet)`
3. `otSysMainloopPoll(&mainloop)`
3.1 `select()`
4. `otSysMainloopProcess(instance, &mainloop);`
4.1 if `FD_ISSET(mIntGpioValueFd, aReadFdSet)` then `read(mIntGpioValueFd, &event, sizeof(event))`

If I understand correctly, whatever happen on `mIntGpioValueFd` will have no effect on `mReadFdSet` between 1. and 2.3
That's the reason why the `CheckInterrupt()` is present in 2.1 to set the timer to 0, else the falling edge of `SPI_INT` is not detected.

And therefore, my initial explanation is coherent.

Do you agree with that?

**Reply by abtink** [reply]
Hi @kimotine ,

Your statement about mainloop and steps is correct. 
But when/how we call `FD_ZERO()` should not matter:
- Whenever there is a transition from HIGH to LOW on SPI_INT pin, its related fd `mIntGpioValueFd` should have an event for us to read.
- This should be tracked all the time, independent of when/if we call `FD_SET()` and/or `FD_ZERO()`.
- `FD_SET()` or `FD_ZERO()` should not change the `fd` itself. 
     - They are simple macros that update a bitmask set which tracks  a set of `fd` (which are `int` values).
- Only `read(mIntGpioValueFd, ...)`  would read and remove the data from `fd`.
- When we call `select()`, as soon as any of the `fd`s in the read fd-set we pass to it, has any data to read `select()` should return.
- So while doing `CheckInterrupt()` works ok (and is fine), it is not technically needed in case 2 (from the top comment) and we could have instead added `mIntGpioValueFd` to read fd-set.
- It should not matter when the GPIO transition happens, before/after/while `FD_SET` or `FD_ZERO` calls or anytime before/while `select()`.

---

One question just to confirm, 
- In your setup, you use the GPIO pin and not use the polling mode, is this correct? 
- I think `gpio-int-device` should be specified in the radio URL for this.

**Reply by kimotine** [reply]
Hi @abtink,
Yes, the GPIO pin is used in polling mode. It is defined in the command line of ot-daemon with `gpio-int-device=/dev/gpiochip5` and 
`gpio-int-line=12`
Most of the time, there is no latency on the SPI_INT
